### PR TITLE
Migrate pkg/kubelet/kubeletconfig to Structured Logging

### PR DESCRIPTION
**Pull request by LittleGoWitch**
#### What type of PR is this?

/kind cleanup
/sig instrumentation
/sig node
/priority important-soon

#### What this PR does / why we need it:
Migrate `pkg/kubelet/kubeletconfig` + `pkg/kubelet/kubelet_getters.go` + `pkg/kubelet/kubelet_resources.go`
to Structured Logging
#### Which issue(s) this PR fixes:

Ref #98976
#### Special notes for your reviewer:
part of [kubernetes/enhancements#1602](https://github.com/kubernetes/enhancements/issues/1602)
As it is related to log, change the log here to structured log.
#### Does this PR introduce a user-facing change?

```release-note
Migrate pkg/kubelet/kubeletconfig+ pkg/kubelet/kubelet_getters.go + pkg/kubelet/kubelet_resources.go
to Structured Logging
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

- [KEP]: [Structured Logging](https://github.com/kubernetes/enhancements/tree/master/keps/sig-instrumentation/1602-structured-logging)

```docs

```

### CODE REVIEW

File: pkg/kubelet/kubeletconfig/util/codec/codec.go [hunk 1]
```diff
@@ -115,7 +115,7 @@ func DecodeKubeletConfiguration(kubeletCodecs *serializer.CodecFactory, data []b
 			return nil, fmt.Errorf("failed lenient decoding: %v", err)
 		}
 		// Continue with the v1beta1 object that was decoded leniently, but emit a warning.
-		klog.Warningf("using lenient decoding as strict decoding failed: %v", err)
+		klog.InfoS("using lenient decoding as strict decoding failed", "err", err)
```
Comment (krzysiekg) [inline pkg/kubelet/kubeletconfig/util/codec/codec.go:118]: Messages should start from a capital letter, to comply with good practices

**Comment by AfrouzMashayekhi** [comment]
Should not we refactor `pkg/kubelet/kubeletconfig/util/log/log.go` as the usage is adding a string to begging of the log. @krzysiekg

**Comment by krzysiekg** [comment]
@AfrouzMashayekhi That's a good question. We definitely have to do something more about this package. I'm not sure what the best course of action is, though.

Option 1 is to add ErrorS and InfoS functions and replace calls to Errorf and Infof with them. There are 40+ calls in this package. These new functions would be even simpler, something like:

```golang
func ErrorS(msg string, keysAndValues ...interface{}) {
	klog.ErrorSDepth(1, fmt.Sprintf(logFmt, msg), keysAndValues...)
}
```

Option 2 is to remove this package altogether and replace it with direct klog calls. Other packages also add prefix and they just duplicate it in every message. One such example is pkg/kubelet/cm/cpumanager/, which adds "[cpumanager]" prefix.

Both approaches seem valid, just a matter of preference in my opinion. Migration instructions don't really enforce anything.

@ehashman @serathius can you recommend anything more specific here?

File: pkg/kubelet/kubeletconfig/configsync.go [hunk 2]
```diff
@@ -198,7 +198,7 @@ func restartForNewConfig(eventClient v1core.EventsGetter, nodeName string, sourc
 	// we directly log and send the event, instead of using the event recorder,
 	// because the event recorder won't flush its queue before we exit (we'd lose the event)
 	event := makeEvent(nodeName, apiv1.EventTypeNormal, KubeletConfigChangedEventReason, message)
-	klog.V(3).Infof("Event(%#v): type: '%v' reason: '%v' %v", event.InvolvedObject, event.Type, event.Reason, event.Message)
+	klog.V(3).InfoS("Event created", "event", klog.KObj(event))
```
Comment (serathius) [inline pkg/kubelet/kubeletconfig/configsync.go:201]: `klog.KObj` is used to reference an object, but it doesn't dump object fields like in this case.

I would propose to add other fields separatly `eventType`, `eventReason`, `eventMessage`, `eventInvolvedObject`, similar to how we pass `podUID` as a separate field. In future we may expand `klog.KObj to put more data about the object, but for now let's have them listed explicitly.

**Comment by serathius** [comment]
I think we should go with option 2, as introducing any shims that just add prefix will break out static analysis that tries to ensure that log calls are correct.

**Comment by AfrouzMashayekhi** [comment]
so should I add `[Kubelet config controller]` or `Kubelet config controller` ,as it is, in the begging of message? @serathius

**Comment by ehashman** [comment]
I agree that we should remove the utility package. Thanks. I will finish a complete review later today.

File: pkg/kubelet/kubeletconfig/checkpoint/download.go [hunk 3]
```diff
@@ -178,25 +178,25 @@ func (r *remoteConfigMap) Download(client clientset.Interface, store cache.Store
 	)
 	// check the in-memory store for the ConfigMap, so we can skip unnecessary downloads
 	if store != nil {
-		utillog.Infof("checking in-memory store for %s", r.APIPath())
+		klog.InfoS("Kubelet config controller checking in-memory store for remoteConfigMap", "remoteConfigMapAPIPath", r.APIPath())
```
Comment (ehashman) [inline pkg/kubelet/kubeletconfig/checkpoint/download.go:181]:
```suggestion
		klog.InfoS("Kubelet config controller checking in-memory store for remoteConfigMap", "apiPath", r.APIPath())
```

Right now each key in these files are using their own relatively verbose prefixes, so none of these columns will align. I think we should try to simply for common patterns here. We already log that this is for the remoteConfigMap so we don't need to repeat that multiple times.

File: pkg/kubelet/kubeletconfig/checkpoint/download.go [hunk 4]
```diff
@@ -178,25 +178,25 @@ func (r *remoteConfigMap) Download(client clientset.Interface, store cache.Store
...
 		cm, err = getConfigMapFromStore(store, r.source.ConfigMap.Namespace, r.source.ConfigMap.Name)
 		if err != nil {
 			// just log the error, we'll attempt a direct download instead
-			utillog.Errorf("failed to check in-memory store for %s, error: %v", r.APIPath(), err)
+			klog.ErrorS(err, "Kubelet config controller failed to check in-memory store for remoteConfigMap", "remoteConfigMapAPIPath", r.APIPath())
```
Comment (ehashman) [inline pkg/kubelet/kubeletconfig/checkpoint/download.go:185]:
```suggestion
			klog.ErrorS(err, "Kubelet config controller failed to check in-memory store for remoteConfigMap", "apiPath", r.APIPath())
```

File: pkg/kubelet/kubeletconfig/checkpoint/download.go [hunk 5]
```diff
@@ -178,25 +178,25 @@ func (r *remoteConfigMap) Download(client clientset.Interface, store cache.Store
...
 		} else if cm != nil {
-			utillog.Infof("found %s in in-memory store, UID: %s, ResourceVersion: %s", r.APIPath(), cm.UID, cm.ResourceVersion)
+			klog.InfoS("Kubelet config controller found remoteConfigMap in in-memory store", "remoteConfigMapAPIPath", r.APIPath(), "remoteConfigMapUID", cm.UID, "remoteConfigMapResourceVersion", cm.ResourceVersion)
```
Comment (ehashman) [inline pkg/kubelet/kubeletconfig/checkpoint/download.go:187]:
```suggestion
			klog.InfoS("Kubelet config controller found remoteConfigMap in in-memory store", "apiPath", r.APIPath(), "remoteConfigMapUID", cm.UID, "resourceVersion", cm.ResourceVersion)
```

File: pkg/kubelet/kubeletconfig/checkpoint/download.go [hunk 6]
```diff
@@ -178,25 +178,25 @@ func (r *remoteConfigMap) Download(client clientset.Interface, store cache.Store
...
 		} else {
-			utillog.Infof("did not find %s in in-memory store", r.APIPath())
+			klog.InfoS("Kubelet config controller did not find remoteConfigMap in in-memory store", "remoteConfigMapAPIPath", r.APIPath())
```
Comment (ehashman) [inline pkg/kubelet/kubeletconfig/checkpoint/download.go:189]:
```suggestion
			klog.InfoS("Kubelet config controller did not find remoteConfigMap in in-memory store", "apiPath", r.APIPath())
```

File: pkg/kubelet/kubeletconfig/checkpoint/download.go [hunk 7]
```diff
@@ -178,25 +178,25 @@ func (r *remoteConfigMap) Download(client clientset.Interface, store cache.Store
...
 		}
 	}
 	// if we didn't find the ConfigMap in the in-memory store, download it from the API server
 	if cm == nil {
-		utillog.Infof("attempting to download %s", r.APIPath())
+		klog.Infof("Kubelet config controller attempting to download remoteConfigMap", "remoteConfigMapAPIPath", r.APIPath())
```
Comment (ehashman) [inline pkg/kubelet/kubeletconfig/checkpoint/download.go:194]:
```suggestion
		klog.Infof("Kubelet config controller attempting to download remoteConfigMap", "apiPath", r.APIPath())
```

File: pkg/kubelet/kubeletconfig/checkpoint/download.go [hunk 8]
```diff
@@ -178,25 +178,25 @@ func (r *remoteConfigMap) Download(client clientset.Interface, store cache.Store
...
 		cm, err = client.CoreV1().ConfigMaps(r.source.ConfigMap.Namespace).Get(context.TODO(), r.source.ConfigMap.Name, metav1.GetOptions{})
 		if err != nil {
 			return nil, status.DownloadError, fmt.Errorf("%s, error: %v", status.DownloadError, err)
 		}
-		utillog.Infof("successfully downloaded %s, UID: %s, ResourceVersion: %s", r.APIPath(), cm.UID, cm.ResourceVersion)
+		klog.InfoS("Kubelet config controller successfully downloaded remoteConfigMap", "remoteConfigMapAPIPath", r.APIPath(), "configMapUID", cm.UID, "configMapResourceVersion", cm.ResourceVersion)
```
Comment (ehashman) [inline pkg/kubelet/kubeletconfig/checkpoint/download.go:199]:
```suggestion
		klog.InfoS("Kubelet config controller successfully downloaded remoteConfigMap", "apiPath", r.APIPath(), "configMapUID", cm.UID, "resourceVersion", cm.ResourceVersion)
```

File: pkg/kubelet/kubeletconfig/checkpoint/store/fsstore.go [hunk 9]
```diff
@@ -113,7 +113,7 @@ func (s *fsStore) Load(source checkpoint.RemoteConfigSource) (*kubeletconfig.Kub
 		return nil, fmt.Errorf("no checkpoint for source %s", sourceFmt)
 	}
 	// load the kubelet config file
-	utillog.Infof("loading Kubelet configuration checkpoint for source %s", sourceFmt)
+	klog.InfoS("Kubelet config controller loading Kubelet configuration checkpoint for source", "sourceAPIPath", source.APIPath(), "sourceUID", source.UID(), "sourceResourceVersion", source.ResourceVersion())
```
Comment (ehashman) [inline pkg/kubelet/kubeletconfig/checkpoint/store/fsstore.go:116]:
```suggestion
	klog.InfoS("Kubelet config controller loading Kubelet configuration checkpoint for source", "apiPath", source.APIPath(), "sourceUID", source.UID(), "resourceVersion", source.ResourceVersion())
```

File: pkg/kubelet/kubeletconfig/configsync.go [hunk 10]
```diff
@@ -156,7 +155,7 @@ func (cc *Controller) saveConfigCheckpoint(source checkpoint.RemoteConfigSource,
 		return status.InternalError, fmt.Errorf("%s, error: %v", status.InternalError, err)
 	}
 	if ok {
-		utillog.Infof("checkpoint already exists for %s, UID: %s, ResourceVersion: %s", source.APIPath(), payload.UID(), payload.ResourceVersion())
+		klog.InfoS("Kubelet config controller checkpoint already exists for source", "sourceAPIPath", source.APIPath(), "payloadUID", payload.UID(), "payloadSourceVersion", payload.ResourceVersion())
```
Comment (ehashman) [inline pkg/kubelet/kubeletconfig/configsync.go:158]:
```suggestion
		klog.InfoS("Kubelet config controller checkpoint already exists for source", "apiPath", source.APIPath(), "payloadUID", payload.UID(), "resourceVersion", payload.ResourceVersion())
```

File: pkg/kubelet/kubeletconfig/configsync.go [hunk 11]
```diff
@@ -198,11 +197,11 @@ func restartForNewConfig(eventClient v1core.EventsGetter, nodeName string, sourc
...
 	// because the event recorder won't flush its queue before we exit (we'd lose the event)
 	event := makeEvent(nodeName, apiv1.EventTypeNormal, KubeletConfigChangedEventReason, message)
-	klog.V(3).Infof("Event(%#v): type: '%v' reason: '%v' %v", event.InvolvedObject, event.Type, event.Reason, event.Message)
+	klog.V(3).InfoS("Event created", "event", klog.KObj(event), "eventInvolvedObject", event.InvolvedObject, "eventType", event.Type, "eventReason", event.Reason, "eventMessage", event.Message)
 	if _, err := eventClient.Events(apiv1.NamespaceDefault).Create(context.TODO(), event, metav1.CreateOptions{}); err != nil {
-		utillog.Errorf("failed to send event, error: %v", err)
+		klog.ErrorS(err, "Kubelet config controller failed to send event")
 	}
-	utillog.Infof(message)
+	klog.InfoS("Kubelet config controller", "eventMessage", message)
```
Comment (ehashman) [inline pkg/kubelet/kubeletconfig/configsync.go:204]:
```suggestion
	klog.InfoS("Kubelet config controller", "message", message)
```

File: pkg/kubelet/kubeletconfig/controller.go [hunk 12]
```diff
@@ -319,6 +319,6 @@ func (cc *Controller) graduateAssignedToLastKnownGood() error {
 	}
 	// update the status to reflect the new last-known-good config
 	cc.configStatus.SetLastKnownGood(assigned.NodeConfigSource())
-	utillog.Infof("updated last-known-good config to %s, UID: %s, ResourceVersion: %s", assigned.APIPath(), assigned.UID(), assigned.ResourceVersion())
+	klog.InfoS("Kubelet config controller updated last-known-good config to new checkpointStore", "newCheckpointStoreAPIPath", assigned.APIPath(), "newCheckpointStoreUID", assigned.UID(), "newCheckpointStoreResourceVersion", assigned.ResourceVersion())
```
Comment (ehashman) [inline pkg/kubelet/kubeletconfig/controller.go:322]:
```suggestion
	klog.InfoS("Kubelet config controller updated last-known-good config to new checkpointStore", "apiPath", assigned.APIPath(), "newCheckpointStoreUID", assigned.UID(), "resourceVersion", assigned.ResourceVersion())
```

**Review by ehashman — CHANGES_REQUESTED** [review]
I'm okay with the log prefixes as written. I think they provide relatively useful context, without them the messages are a little confusing.

File: pkg/kubelet/kubeletconfig/checkpoint/download.go [hunk 13]
```diff
@@ -178,25 +178,25 @@ func (r *remoteConfigMap) Download(client clientset.Interface, store cache.Store
...
-		utillog.Infof("checking in-memory store for %s", r.APIPath())
+		klog.InfoS("Kubelet config controller checking in-memory store for remoteConfigMap", "apiPath", r.APIPath())
 		cm, err = getConfigMapFromStore(store, r.source.ConfigMap.Namespace, r.source.ConfigMap.Name)
 		if err != nil {
 			// just log the error, we'll attempt a direct download instead
-			utillog.Errorf("failed to check in-memory store for %s, error: %v", r.APIPath(), err)
+			klog.ErrorS(err, "Kubelet config controller failed to check in-memory store for remoteConfigMap", "apiPath", r.APIPath())
 		} else if cm != nil {
-			utillog.Infof("found %s in in-memory store, UID: %s, ResourceVersion: %s", r.APIPath(), cm.UID, cm.ResourceVersion)
+			klog.InfoS("Kubelet config controller found remoteConfigMap in in-memory store", "apiPath", r.APIPath(), "remoteConfigMapUID", cm.UID, "resourceVersion", cm.ResourceVersion)
```
Comment (serathius) [inline pkg/kubelet/kubeletconfig/checkpoint/download.go:187]:
```suggestion
			klog.InfoS("Kubelet config controller found remote ConfigMap in in-memory store", "apiPath", r.APIPath(), "configMapUID", cm.UID, "resourceVersion", cm.ResourceVersion)
```

File: pkg/kubelet/kubeletconfig/configsync.go [hunk 14]
```diff
@@ -156,7 +155,7 @@ func (cc *Controller) saveConfigCheckpoint(source checkpoint.RemoteConfigSource,
 		return status.InternalError, fmt.Errorf("%s, error: %v", status.InternalError, err)
 	}
 	if ok {
-		utillog.Infof("checkpoint already exists for %s, UID: %s, ResourceVersion: %s", source.APIPath(), payload.UID(), payload.ResourceVersion())
+		klog.InfoS("Kubelet config controller checkpoint already exists for source", "apiPath", source.APIPath(), "payloadUID", payload.UID(), "resourceVersion", payload.ResourceVersion())
```
Comment (serathius) [inline pkg/kubelet/kubeletconfig/configsync.go:158]:
```suggestion
		klog.InfoS("Kubelet config controller checkpoint already exists for source", "apiPath", source.APIPath(), "checkpointUID", payload.UID(), "resourceVersion", payload.ResourceVersion())
```

File: pkg/kubelet/kubeletconfig/configsync.go [hunk 15]
```diff
@@ -198,11 +197,11 @@ func restartForNewConfig(eventClient v1core.EventsGetter, nodeName string, sourc
 	// we directly log and send the event, instead of using the event recorder,
 	// because the event recorder won't flush its queue before we exit (we'd lose the event)
 	event := makeEvent(nodeName, apiv1.EventTypeNormal, KubeletConfigChangedEventReason, message)
-	klog.V(3).Infof("Event(%#v): type: '%v' reason: '%v' %v", event.InvolvedObject, event.Type, event.Reason, event.Message)
+	klog.V(3).InfoS("Event created", "event", event, "involvedObject", event.InvolvedObject, "eventType", event.Type, "reason", event.Reason, "message", event.Message)
```
Comment (serathius) [inline pkg/kubelet/kubeletconfig/configsync.go:200]:
```suggestion
	klog.V(3).InfoS("Event created", "event", klog.KObj(event), "eventInvolvedObject", klog.KObj(event.InvolvedObject), "eventType", event.Type, "eventReason", event.Reason, "eventMessage", event.Message)
```

File: pkg/kubelet/kubeletconfig/configsync.go [hunk 16]
```diff
@@ -198,11 +197,11 @@ func restartForNewConfig(eventClient v1core.EventsGetter, nodeName string, sourc
...
 	if _, err := eventClient.Events(apiv1.NamespaceDefault).Create(context.TODO(), event, metav1.CreateOptions{}); err != nil {
-		utillog.Errorf("failed to send event, error: %v", err)
+		klog.ErrorS(err, "Kubelet config controller failed to send event")
 	}
-	utillog.Infof(message)
+	klog.InfoS("Kubelet config controller", "message", message)
```
Comment (serathius) [inline pkg/kubelet/kubeletconfig/configsync.go:204]:
```suggestion
	klog.InfoS("Kubelet config controller", "eventMessage", message)
```

File: pkg/kubelet/kubeletconfig/controller.go [hunk 17]
```diff
@@ -319,6 +319,6 @@ func (cc *Controller) graduateAssignedToLastKnownGood() error {
 	}
 	// update the status to reflect the new last-known-good config
 	cc.configStatus.SetLastKnownGood(assigned.NodeConfigSource())
-	utillog.Infof("updated last-known-good config to %s, UID: %s, ResourceVersion: %s", assigned.APIPath(), assigned.UID(), assigned.ResourceVersion())
+	klog.InfoS("Kubelet config controller updated last-known-good config to new checkpointStore", "apiPath", assigned.APIPath(), "newCheckpointStoreUID", assigned.UID(), "resourceVersion", assigned.ResourceVersion())
```
Comment (serathius) [inline pkg/kubelet/kubeletconfig/controller.go:322]:
```suggestion
	klog.InfoS("Kubelet config controller updated last-known-good config to new checkpointStore", "apiPath", assigned.APIPath(), "checkpointUID", assigned.UID(), "resourceVersion", assigned.ResourceVersion())
```

File: pkg/kubelet/kubeletconfig/configsync.go (hunk 11 above)
Comment (AfrouzMashayekhi) [reply pkg/kubelet/kubeletconfig/configsync.go]: @serathius Elana suggested this change. so which one? :D

File: pkg/kubelet/kubeletconfig/configsync.go [hunk 18]
```diff
@@ -198,11 +197,11 @@ func restartForNewConfig(eventClient v1core.EventsGetter, nodeName string, sourc
 	// we directly log and send the event, instead of using the event recorder,
 	// because the event recorder won't flush its queue before we exit (we'd lose the event)
 	event := makeEvent(nodeName, apiv1.EventTypeNormal, KubeletConfigChangedEventReason, message)
-	klog.V(3).Infof("Event(%#v): type: '%v' reason: '%v' %v", event.InvolvedObject, event.Type, event.Reason, event.Message)
+	klog.V(3).InfoS("Event created", "event", klog.KObj(event), "eventInvolvedObject", event.InvolvedObject, "eventType", event.Type, "eventReason", event.Reason, "eventMessage", event.Message)
```
[the remark this answers is not in the document]
Comment (AfrouzMashayekhi) [reply pkg/kubelet/kubeletconfig/configsync.go]: @serathius this one too

**Review by ehashman — COMMENTED** [review]
Addressed in #100265

**Comment by ehashman** [comment]
Closing in favour of #100265 since no response. Don't worry, I kept your commit authorship so you will get credit :)

/close

**Pull request closed without merging**